### PR TITLE
ENH: add IIR comb filter to scipy.signal

### LINKED ISSUE #11558: IIR comb filter
**Describe the solution you'd like**
I think it would be very useful to support the design of IIR comb filters. 

**Describe alternatives you've considered**
The only alternative I could think of would be the use of multiple IIR notch filters, but doesn't seem like a great idea.

### PR DESCRIPTION
**Pull request by skresearcher**
#### Reference issue
Closes #11558 

#### What does this implement/fix?
Implements IIR notching and peaking comb filters in the scipy.signal module.

### CODE REVIEW

File: scipy/signal/tests/test_filter_design.py
Comment (larsoner) [inline scipy/signal/tests/test_filter_design.py:3621]: These need to each be in their own context manager for them all to execute (the first failure will kick you out of the context without executing the subsequent lines). You could for example loop over
```
for args in [(-fs, 30, 4), ...]:
    with pytest.raises(...):
        iircomp(*args, fs=fs)
```

File: scipy/signal/tests/test_filter_design.py
Comment (larsoner) [inline scipy/signal/tests/test_filter_design.py:3640]: nowadays we can `@pytest.parametrize('ftypes', ('notch', 'peak')` the `test_frequency_response(self, ftypes)` function

File: scipy/signal/tests/test_filter_design.py
Comment (larsoner) [inline scipy/signal/tests/test_filter_design.py:3656]: The SciPy codebase prefers using the numpy asserts like `assert_allclose` for stuff like this (no need / advantage to introduce a new convention for people to learn here AFAIK)

File: scipy/signal/tests/test_filter_design.py
Comment (larsoner) [inline scipy/signal/tests/test_filter_design.py:3669]: pytest directly handles `assert` with an informative traceback so you can just do `assert issubclass(...)`

File: scipy/signal/filter_design.py
Comment (larsoner) [inline scipy/signal/filter_design.py:4917]: There is no reference to `[1]_` above, might be worth adding a quick `Notes` section to say "For implementation details, see [1]_."

Also, typically for high-order IIR filters there can be issues with stability when represented in TF (b, a) form, hence why we have second-order-sections filtering. It might be worth mentioning that this is not the case for these filters, and why (a quick look suggests it's just because there is a single repeated pole, which won't suffer from precision loss like high-order filters with non-repeated poles will...?)

File: scipy/signal/tests/test_filter_design.py
Comment (larsoner) [inline scipy/signal/tests/test_filter_design.py:3655]: Does it make sense to also add a test that the comb and notch are inverses (combining them approximately yields a delta function / no-op)? I think it would probably work, assuming the numerical values don't become problematic somehow.

File: scipy/signal/tests/test_filter_design.py
Comment (skresearcher) [reply scipy/signal/tests/test_filter_design.py]: Fixed.

File: scipy/signal/tests/test_filter_design.py
Comment (skresearcher) [reply scipy/signal/tests/test_filter_design.py]: Fixed.

File: scipy/signal/tests/test_filter_design.py
Comment (skresearcher) [reply scipy/signal/tests/test_filter_design.py]: Fixed.

File: scipy/signal/tests/test_filter_design.py
Comment (skresearcher) [reply scipy/signal/tests/test_filter_design.py]: Fixed.

File: scipy/signal/filter_design.py
Comment (skresearcher) [reply scipy/signal/filter_design.py:4923]: Fixed.

File: scipy/signal/tests/test_filter_design.py
Comment (skresearcher) [reply scipy/signal/tests/test_filter_design.py]: I tried what you said, but the result is outputting an unusual "peak/notch" hybrid. Below are the links to the code I tested for this scenario and the corresponding plot. 

https://github.com/skresearcher/audiotools/blob/master/iircomb-notchandpeak.py
https://github.com/skresearcher/audiotools/blob/master/iircomb-notchandpeak.png

**Comment by larsoner** [comment]
The combined filter is the product of the two not the sum (computing it here two ways -- mult of the freq responses and also by polynomial-multiplicatiion-by-convolution-of-coefficients of the Z transforms then computing the freq resp):

<details>

```Python
import numpy as np
from scipy import signal
import matplotlib.pyplot as plt

b_notch, a_notch = signal.iircomb(60, 35, 10, ftype='notch', fs=600)
b_peak, a_peak = signal.iircomb(60, 35, 10, ftype='peak', fs=600)
w_notch, h_notch = signal.freqz(b_notch, a_notch, fs=600)
w_peak, h_peak = signal.freqz(b_peak, a_peak, fs=600)

b_combined = np.convolve(b_notch, b_peak)
a_combined = np.convolve(a_notch, a_peak)
w_combined, h_combined = signal.freqz(b_combined, a_combined, fs=600)
h_combined_fd = h_notch * h_peak
plt.plot(w_notch, abs(h_notch), label='notch')
plt.plot(w_peak, abs(h_peak), label='peak')
plt.plot(w_combined, abs(h_combined), '--', label='combined')
plt.plot(w_combined, abs(h_combined), ':', label='combined (freq)')
plt.legend(bbox_to_anchor=(1.05, 1))
plt.tight_layout()
```

</details>

Figure_1

But there is something very strange going on here. Changing the first argument from 60 to, say, 50, which I expected to change where the notches occurred (from 60 Hz plus harmonics down to 50 Hz plus harmonics) we get almost the exact same thing:

Figure_1


It seems like you actually control the notch frequencies by changing the order `N` -- If I change it to 12, I get a notch at 50 Hz:

Figure_2


This seems consistent with the MATLAB implementation and description, where you actually give the number of notches, not the desired frequency:

https://www.mathworks.com/help/dsp/ref/iircomb.html

So there is something weird about the interface here. 

I'm also not sure why `N` is limited to 24. Is there a reason for this? If I remove the check and set `N=30`, I get a notch filter at 20 Hz plus harmonics, which is what I'd expect (based on the MATLAB docs, not the docs here):

Figure_3

And if I use MATLAB and design a peak and notch filter:

<details>

```MATLAB
>> [b_notch, a_notch] = iircomb(10, 0.1, 'notch');
>> [b_peak, a_peak] = iircomb(10, 0.01, 'peak');
>> [h_notch, w_notch] = freqz(b_notch, a_notch, 300, 600);
>> [h_peak, w_peak] = freqz(b_peak, a_peak, 300, 600);
>> plot(w_notch, abs(h_notch))
>> hold on
>> plot(w_notch, abs(h_notch))
```

</details>

I get:

untitled

So indeed these are not meant to be inverses of one another. Maybe this is obvious to people who have used IIR comb filters before (I haven't actually!). But in any case, I find MATLAB's interface involving `N` to be clearer, because you can't actually notch at arbitrary frequencies like the interface here seems to suggest.

If we want to keep the interface based on fundamental frequency of the notch to remove (which on the face of it does seem cleaner), we probable need to leave `N` out of the equation as a parameter, and compute it from `fs` and the specified frequency `w0`, probably raising some error if it's not sufficiently close to something that can actually be comb filtered (?), i.e., produces an integer value of `N` in the computation (30, 50, and 60 Hz are all valid choices because they evenly divide into 600, but something like 57 is not because it does not evenly divide into 600).

Maybe I'm misunderstanding how it's meant to work from the docstring, but if you want to remove a specific frequency (and its harmonics) from the signal, it seems like you need to compute `N` (and the frequency needs to evenly divide into `fs`)...

**Comment by skresearcher** [comment]
> The combined filter is the product of the two not the sum (computing it here two ways -- mult of the freq responses and also by polynomial-multiplicatiion-by-convolution-of-coefficients of the Z transforms then computing the freq resp):

Thanks for clarifying this and doing the detailed testing. I can see that the comb filter isn't working in some cases because `N` is currently independent of `w0` and `fs`.

> I find MATLAB's interface involving N to be clearer, because you can't actually notch at arbitrary frequencies like the interface here seems to suggest.

I used the argument interface from SciPy's `iirnotch` and `iirpeak` functions with `w0` and `fs`. I wanted to keep it similar to other SciPy filter functions where the cutoff frequency is the primary input.  

> If we want to keep the interface based on fundamental frequency of the notch to remove (which on the face of it does seem cleaner), we probable need to leave N out of the equation as a parameter, and compute it from fs and the specified frequency w0. probably raising some error if it's not sufficiently close to something that can actually be comb filtered (?), i.e., produces an integer value of N in the computation (30, 50, and 60 Hz are all valid choices because they evenly divide into 600, but something like 57 is not because it does not evenly divide into 600).

I tested this approach, and it works. I would prefer this, unless you think it is better to keep it similar to MATLAB's interface with just `N`. My only question is whether we should raise a warning or throw an exception if `fs/w0` does not divide evenly.

> I'm also not sure why N is limited to 24. Is there a reason for this? If I remove the check and set N=30, I get a notch filter at 20 Hz plus harmonics, which is what I'd expect (based on the MATLAB docs, not the docs here):

I had issues with overflowing when working on the gammatone filters, so I thought I could apply the same check here. Now that `N` is not going to be an input argument and instead computed based on `w0` and `fs` I can remove that restriction.

**Comment by larsoner** [comment]
> My only question is whether we should raise a warning or throw an exception if fs/w0 does not divide evenly.

It's safer to raise an exception. People can adjust their code to compute and use the nearest one if they actually want that behavior.

**Comment by skresearcher** [comment]
I've made the changes.

**Comment by skresearcher** [comment]
@larsoner if there are no issues with this PR, can we merge it?

**Comment by e-q** [comment]
Bit of a drive-by comment, but it would be nice to have more output options than `b, a` form. Wherever possible, I think it would be good to have ZPK and SOS output options for filter design functions.

**Comment by skresearcher** [comment]
@e-q I provided ZPK and SOS output options for the gammatone filters that I recently contributed, but after discussion we decided to just return `b, a` form as users can use the `tf2zpk` and `tf2sos` functions to convert easily between different filter formats.

**Comment by skresearcher** [comment]
@larsoner this PR hasn't been merged yet. If there are no issues, can you please merge it?

**Comment by e-q** [comment]
@skresearcher Unfortunately, I missed that discussion. Conversion from TF to ZPK after the fact can be imprecise due to root finding with finite precision polynomial coefficients, especially for higher order or compound filters, making it tougher to calculate a precise frequency response. Furthermore, this makes SOS conversion imprecise,  As tf2sos just calls zpk2sos on tf2zpk.

**Comment by larsoner** [comment]
> Conversion from TF to ZPK after the fact can be imprecise due to root finding with finite precision polynomial coefficients, especially for higher order or compound filters, making it tougher to calculate a precise frequency response. Furthermore, this makes SOS conversion imprecise, As tf2sos just calls zpk2sos on tf2zpk.

One consideration that went into the decision was that the advantages of ZPK -> SOS + sosfilt come from avoiding unstable numerical computations (e.g., when there are poles close to one another and the unit circle) -- I don't think that should be the case for gammatone filters or this PR. Though admittedly I have not checked, I'm just assuming this is the case from looking at the filters...

The second consideration was that the formulas for the gammatone filters themselves were TF / they were constructed in that space. If someday we instead construct them directly from poles and zeros, I agree at that point it would make sense to support an `output` argument and internally always construct ZPK and only convert to TF output when requested (though this would maybe be by default). As it stands, any degradation has already occurred. Same thing here -- the filter denominator is just directly computed as `1 - az^-N or 1 + az^-N` (and similarly for the numerator).

So I'm +1 or just returning these in TF form for now. It stays truer to the underlying implementation, and won't mislead people into thinking that requesting a ZPK or SOS output will help shield them from precision loss (which it does for most of our other filter functions, hence why it makes sense to have the output argument there).

**Comment by ilayn** [comment]
We should be able to get the ZPKs carefully from those num/den pairs though.

**Comment by larsoner** [comment]
> We should be able to get the ZPKs carefully from those num/den pairs though.

I'm happy to leave this as a subsequent PR once someone has a use case for it. Maybe our ZPK conversion routines are already good enough, maybe people won't really need these in ZPK form, etc. I'd rather not hold up this PR for this since we can always add it later.

**Comment by e-q** [comment]
> The second consideration was that the formulas for the gammatone filters themselves were TF / they were constructed in that space.

Ah, I overlooked that. That makes total sense, and the note in the docstring that @skresearcher includes is helpful for the user. As @ilayn suggests, it would be nice to include a careful calculation of the filter's zeros and poles if the order is high, but I agree that this shouldn't block this PR.

**Comment by ilayn** [comment]
same here, I am perfectly fine to do it some other PR

File: scipy/signal/filter_design.py
Comment (larsoner) [inline scipy/signal/filter_design.py:4883]: I think we need to rework the wording here. Something like "frequency spacing to use". Because the notch filter will filter at 0, w0, 2*w0, ..., but the peak filter will put peaks at 0.5*w0, 1.5*w0. Can you think of a better wording?

Comment (skresearcher) [reply scipy/signal/filter_design.py:4883]: I'm thinking of including the following under the `ftype` parameter, if that's okay. 

"If 'notch', function returns filter with notches at frequencies `0`, `w0`, `2 * w0`, ... If 'peak', function returns filter with peaks at frequencies `0.5 * w0`, `1.5 * w0`, `2.5 * w0`, ..."

If not, I can modify the `w0` parameter description like you've mentioned.

Comment (larsoner) [reply scipy/signal/filter_design.py:4883]: Modifying `ftype` is fine by me

Comment (skresearcher) [reply scipy/signal/filter_design.py]: Fixed.

**Review by larsoner — APPROVED** [review]
LGTM, will merge once CIs come back happy since @ilayn and @e-q are also satisfied

**Comment by larsoner** [comment]
Thanks @skresearcher !

**Pull request merged**
